ssh_proxy_server/forwarders/sftp.py

```diff
@@ -67,3 +70,42 @@ def write(self, offset, data):
         data = self.plugin.handle_data(data)
         self.writefile.write(data)
         return paramiko.SFTP_OK
+
+
+class SFTPHandlerReplacePlugin(SFTPHandlerPlugin):
+    """
+    Replaces a SFTP transmitted File during transit
+    """
+    @classmethod
+    def parser_arguments(cls):
+        cls.PARSER.add_argument(
+            '--sftp-replace',
+            dest='sftp_replacement_file',
+            required=True,
+            help='file that is used for replacement'
+        )
+
+    def __init__(self, sftp, filename):
+        super().__init__(sftp, filename)
+        logging.info("sftp file transfer detected: %s", filename)
+        logging.info("intercepting sftp file, replacement: %s", self.args.sftp_replacement_file)
+        self.file = None
+        self.replacement = open(self.args.sftp_replacement_file, "rb")
+        self.buf_size = None
+
+    def close(self):
+        self.replacement.close()
+
+    def handle_data(self, data):
+        # Cannot reasonably detect given buffer size, using best guess: biggest power of 2
```

Comment:
Why are you using this workarround with biggest power of 2?

This method is thread safe, and your file descriptor should remeber its current position. So you can read the same amount of data as given from the remote client/server

Reply:
Yes, in theory, that was the first thing I tried and tested, but for some reason the SFTP Client on the host will reject and close the SFTP channel if it receives more than an arbitrary amount of data.
An error that occurs mainly with small files. For example, when passing through data from the replacement file with a buffer size as big as the amount of data that was requested at the initial transferring of the file, following error message will be thrown:

`Received more data than asked for 141 > 56`  

Keep in mind that the file has a size of just 141 bytes.

Reply:
small files and files up to the size of the wanted files are transfered without errors. Only larger files does not work.
Tested with commit: 598e2d22eb6c4543a4942dd6992d359b4a4ab993